### DISCUSSION TITLE
Future-proof simplify() for more than two methods

### DISCUSSION THREAD

**Discussion by mwtoews** [shapely/shapely#1490] [Ideas]
In the past, GEOS had two simplify methods:
- simplifyDP for the [Douglas–Peucker algorithm](https://en.wikipedia.org/wiki/Ramer%E2%80%93Douglas%E2%80%93Peucker_algorithm) using a distance threshold.
- simplifyTP similar to above, but for `preserve_topology = True`. This is shapely's default method.

With only two methods, the signature `simplify(self/geometry, tolerance, preserve_topology=True)` is fine.

However, there are a plethora of different simplify algorithms (i.e. #1046). For instance, JTS has simplifyVW for the [Visvalingam–Whyatt algorithm](https://en.wikipedia.org/wiki/Visvalingam%E2%80%93Whyatt_algorithm), which is not yet available in GEOS, but is ticketed in https://github.com/libgeos/geos/issues/635 to be ported (other simplify methods are in GEOS' [src/simplify](https://github.com/libgeos/geos/tree/848bce44d308a05d082766da21ea5d9db2e79b8e/src/simplify) directory.)  And there is further interest with other methods, e.g. #1360. With time it is inevitable there will be more than two simplify methods.

I suggest that the signature to be changed to something like: `simplify(self/geometry, tolerance, method="TP", **kwargs)`. The keyword `method` is used to choose an available method (currently two). The value could be either a str or possibly an Enum. And `**kwargs` can be used to handle backwards compatibility with `preserve_topology`, or if other parameters are needed by other methods. Discuss!

**Comment by jorisvandenbossche** [comment]
+1 on the idea!

I had one concern about that something like `preserve_topology=True` is more accessible than `method="TP"` (although we can also make it `method="topology_preserving"` in full to be clearer), but since that is now the default, you only have to specify `method` (and dive into the names and explanations to see which one you need) if you want one of the other methods. So that's not really a concern.

Sidenote, the GEOS docs mention that both GEOSSimplify and GEOSTopologyPreserveSimplify are using the Douglas/Peucker algorithm (although that's maybe a copy/paste leftover from adding docstrings)

**Reply by mwtoews** [reply]
The [TopologyPreservingSimplifier](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/simplify/TopologyPreservingSimplifier.html) "uses a maximum-distance difference algorithm similar to the Douglas-Peucker algorithm". And yes, please discuss ideas on (e.g.) what values "method" should expect.

**Reply by mwtoews** [reply]
It wouldn't be too difficult to document a set of expected keyword strings, but then raise an exception describing "method 'X' is not understood, choose one of 'A', 'B', or 'C'". Internally, a function would be used to validate the "method" keyword, which could do some loose matching, e.g. anything with "TP" or "topology" would refer to GEOSTopologyPreserveSimplify, etc. This same function would also need to check the GEOS version to ensure that the chosen method (e.g.) simplifyVW is available.

**Comment by brendan-ward** [comment]
Given that each simplification method may have different parameters, I wonder if it would be better to split these into separate functions, rather than using a `method` keyword which then interacts with other keywords:

* `simplify_dp()`
* `simplify()` alias for `simplify_dp()` but consider deprecating the bare `simplify` in a future release
* `simplify_vw()` once available from GEOS
* other simplification methods get their own functions

This would also make it easier for the docstring of each to more clearly describe the algorithm and tradeoffs.

**Reply by jorisvandenbossche** [reply]
If they would have different parameters, that could indeed be a good reason to have separate methods. 
But if that would _not_ be the case (the two methods currently supported both have the same tolerance keyword, and the [VW simplifier in JTS](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/simplify/VWSimplifier.html) also has a tolerance keyword), I think I would prefer a single method with a `method` keyword. That would avoid a proliferation of functions ànd equivalent methods on the geometry classes (and methods on geopandas objects). And I actually think that a single docstring about the different methods can maybe even do a better job at describing the tradeoffs to choose between the different methods (of course we could also have a good user guide page about that). 

Of course, that's difficult to predict for potential future methods, and then separate functions might be more future proof.

**Reply by mwtoews** [reply]
Thanks for the suggestion! I did consider suggesting to add separate methods. There are a few other things to consider with this approach. First, each simplify method would need to be added to both the main namespace ([`shapely.simplify`](https://github.com/shapely/shapely/blob/03284740014ba0d8b1917752a6b5dcbfd4ffeb39/shapely/constructive.py#L737-L776)) and the [base geometry class](https://github.com/shapely/shapely/blob/03284740014ba0d8b1917752a6b5dcbfd4ffeb39/shapely/geometry/base.py#L452-L461). If the method were not available due to shapely version, it would raise AttributeError to the user. Also, most simplify methods that I've seen just have a single "tolerance" parameter. [GEOSPolygonHullSimplify](https://github.com/libgeos/geos/blob/848bce44d308a05d082766da21ea5d9db2e79b8e/capi/geos_c.h.in#L3502-L3522) takes different parameters. [Reduce bend](https://github.com/NRCan/geo_sim_processing#reduce-bend) ([code](https://github.com/NRCan/geo_sim_processing/blob/master/reduce_bend_algorithm.py)) takes "tolerance" and a few other parameters. These additional parameters could be consumed with `simplify(... **kwargs)`, raising exceptions if other parameters are expected.
